internal/client/api_clientbundle.go

```diff
@@ -4,110 +4,159 @@ import (
 	"archive/zip"
 	"bytes"
 	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
-	"io/ioutil"
+	"io"
 	"net/http"
 	"strings"
 )
 
 const (
-	URLTargetForClientBundle = "api/clientbundle"
-
-	filenameCAPem      = "ca.pem"
-	filenameCertPem    = "cert.pem"
-	filenamePrivKeyPem = "key.pem"
-	filenamePubKeyPem  = "cert.pub"
-	filenameKubeconfig = "kube.yml"
+	URLTargetForClientBundle           = "api/clientbundle"
+	URLTargetForClientBundleQueryLabel = "label"
+
+	filenameCAPem             = "ca.pem"
+	filenameCertPem           = "cert.pem"
+	filenamePrivKeyPem        = "key.pem"
+	filenamePubKeyPem         = "cert.pub"
+	filenameKubeconfig        = "kube.yml"
+	filenameDockerBundleZip   = "ucp-docker-bundle.zip"
+	filenameDockerBundlerMeta = "meta.json"
 )
 
 var (
 	ErrFailedToRetrieveClientBundle         = errors.New("failed to retrieve the client bundle from MKE")
 	ErrFailedToFindClientBundleMKEPublicKey = errors.New("no MKE Public key was found that matches the client bundle")
 )
 
-// ApiClientBundle retrieve a client bundle(
-func (c *Client) ApiClientBundleCreate(ctx context.Context) (ClientBundle, error) {
+// ApiClientBundle retrieve a client bundle.
+func (c *Client) ApiClientBundleCreate(ctx context.Context, label string) (ClientBundle, error) {
 	var cb ClientBundle
 
 	req, err := c.RequestFromTargetAndBytesBody(ctx, http.MethodPost, URLTargetForClientBundle, []byte{})
 	if err != nil {
 		return cb, err
 	}
 
+	// @SEE https://github.com/Mirantis/orca/blob/master/controller/api/bundle.go#L35 for source.
+	q := req.URL.Query()
+	q.Add(URLTargetForClientBundleQueryLabel, label)
+	req.URL.RawQuery = q.Encode()
+
 	resp, err := c.doAuthorizedRequest(req)
 	if err != nil {
 		return cb, err
 	}
 	defer resp.Body.Close()
 
-	zipBytes, err := ioutil.ReadAll(resp.Body)
+	zb, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return cb, err
 	}
 
-	zipReader, err := zip.NewReader(bytes.NewReader(zipBytes), resp.ContentLength)
+	zr, err := zip.NewReader(bytes.NewReader(zb), resp.ContentLength)
 	if err != nil {
 		return cb, err
 	}
 
-	cb.ID = zipReader.Comment
+	cb.ID = zr.Comment
 
 	errs := []error{}
 
-	for _, f := range zipReader.File {
+	for _, f := range zr.File {
```

Comment:
these for loop + case + if cases looks complex, I wonder if there is a way to simplify ?

Reply:
We are looping over the files in the client bundle zip. This zip contains another zip with more files.
The loop is over a list, and checking properties of each file, so we can't directly look for each known file, without creating more loops (one per file.)
It is poop, but I tried to keep all of the poop in one pile.

Reply:
took another look at it. I don't see the `if` you are referring to? There is a loop+switch over the files in the zip, with a loop+switch for the inner zip file.

Reply:
ah, there is a missing file.Close(). I will add it now.

Reply:
all files are closed now.